### PR TITLE
fix(terminal): report terminal window size in pixels

### LINKED ISSUE #8259: "Terminal does not support reporting screen size s via the TIOCGWINSZ ioctl"
- `nvim --version`: NVIM v0.2.3-975-g84bac9f50
- Vim (version: ) behaves differently?  Different error but doesn't work either
- Operating system/version: macos 10.13
- Terminal name/version: kitty
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

Requires kitty terminal emulator and [its `icat` script](https://gist.github.com/abd37a3d7a0bb3f6f4d8ca15c6b1dfa3)

```
nvim -u NORC
:term
icat (some img file)
```

### Actual behavior

`Terminal does not support reporting screen sizes via the TIOCGWINSZ ioctl`

### Expected behavior

An image should be displayed.

### PR DESCRIPTION
**Pull request by lumynou5**
Window size in pixels is required by some programs, reporting it can help those programs work inside NeoVim terminals.

Closes #8259.

However, this patch doesn't make [`icat`](https://sw.kovidgoyal.net/kitty/kittens/icat/) work. Instead, it just makes the program show another error, because the problem is not only related to window sizes in pixels but how escape sequences (specifically, APC) are handled.  See #4349, #20176.

### CODE REVIEW

File: src/nvim/os/pty_process_unix.c [hunk 1]
```diff
@@ -171,7 +171,13 @@ int pty_process_spawn(PtyProcess *ptyproc)
   Process *proc = (Process *)ptyproc;
   assert(proc->err.closed);
   uv_signal_start(&proc->loop->children_watcher, chld_handler, SIGCHLD);
-  ptyproc->winsize = (struct winsize){ ptyproc->height, ptyproc->width, 0, 0 };
+  ioctl(1, TIOCGWINSZ, &ptyproc->winsize);
+  ptyproc->winsize = (struct winsize){
+    ptyproc->height,
+    ptyproc->width,
+    ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col * ptyproc->width,
+    ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row * ptyproc->height,
```
Comment (justinmk) [inline src/nvim/os/pty_process_unix.c:179]: use parens to make the order of operations obvious
```suggestion
    (ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col) * ptyproc->width,
    (ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row) * ptyproc->height,
```

**Review by justinmk — APPROVED** [review]
Nice, thanks. If no one else has comments, ping me if this isn't merged in a couple days

**Comment by justinmk** [comment]
I wonder if this fixes any of the log messages that were showing in Windows CI https://github.com/neovim/neovim/blob/cf9f002f31c8b4d9d42912a3f45f5d3db4462fd9/test/testutil.lua#L95

**Comment by lumynou5** [comment]
Currently the test is failed.  When stdout is redirected, `ioctl` calls fail with errno 25, and the program crashes due to invalid division.
Adding a check to set pixel sizes to `0` in case of failed calls can help, but it seems not a desirable solution?

```c
  int res = ioctl(1, TIOCGWINSZ, &ptyproc->winsize);
  ptyproc->winsize = (struct winsize){
    ptyproc->height,
    ptyproc->width,
    res ? 0 : (ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col) * ptyproc->width,
    res ? 0 : (ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row) * ptyproc->height,
  };
```

**Comment by zeertzjq** [comment]
It seems this doesn't actually solve anything, and may not be correct. For example, when running Nvim with `--headless` and attaching another TUI with `--remote-ui`, this will use the pixel size of the terminal running the `--headless` instance instead of the one running the `--remote-ui` instance.

**Comment by lumynou5** [comment]
> It seems this doesn't actually solve anything, and may be incorrect. For example, when running Nvim with `--headless` and attaching another TUI with `--remote-ui`, this will use the pixel size of the terminal running the `--headless` instance instead of the one running the `--remote-ui` instance.

Thanks. I didn't think about remote. I'll try to solve it.

**Pull request closed without merging**